src/Gaffer/Collect.cpp

```diff
@@ -0,0 +1,658 @@
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "Gaffer/Collect.h"
+
+#include "Gaffer/NumericPlug.h"
+#include "Gaffer/CompoundNumericPlug.h"
+#include "Gaffer/PlugAlgo.h"
+#include "Gaffer/TypedObjectPlug.h"
+
+#include "IECore/DataAlgo.h"
+#include "IECore/TypeTraits.h"
+
+#include "boost/bind.hpp"
+
+#include "tbb/blocked_range.h"
+#include "tbb/parallel_for.h"
+#include "tbb/parallel_reduce.h"
+
+#include "fmt/format.h"
+
+// using namespace Imath;
+
+using namespace std;
+using namespace tbb;
+using namespace IECore;
+using namespace Gaffer;
+
+namespace
+{
+
+// Miscellaneous utilities
+// =======================
+
+template<typename T>
+void filterVector( vector<T> &v, const vector<unsigned char> &f )
+{
+	size_t newIndex = 0;
+	for( size_t oldIndex = 0; oldIndex < v.size(); ++oldIndex )
+	{
+		if( f[oldIndex] )
+		{
+			if( newIndex != oldIndex )
+			{
+				v[newIndex] = std::move( v[oldIndex] );
+			}
+			++newIndex;
+		}
+	}
+	v.resize( newIndex );
+	v.shrink_to_fit();
+}
+
+void filterVectors( CompoundObject *compoundObject, const vector<unsigned char> &filter )
+{
+	for( auto &[name, value] : compoundObject->members() )
+	{
+		if( auto o = runTimeCast<ObjectVector>( value.get() ) )
+		{
+			filterVector( o->members(), filter );
+		}
+		else
+		{
+			if( auto d = runTimeCast<UCharVectorData>( value.get() ) )
+			{
+				// See `OutputTraits<BoolPlug>`.
+				value = new BoolVectorData( vector<bool>( d->readable().begin(), d->readable().end() ) );
+			}
+
+			IECore::dispatch(
+				static_cast<Data *>( value.get() ),
+				[&] ( auto *data ) {
+					using DataType = remove_pointer_t<decltype( data )>;
+					if constexpr ( TypeTraits::IsVectorTypedData<DataType>::value )
+					{
+						filterVector( data->writable(), filter );
+					}
+				}
+			);
+		}
+	}
+}
+
+using IntRange = blocked_range<int>;
+const IECore::InternedString g_enabledOutputs( "__enabledOutputs__" );
+
+// Type-based plug dispatch
+// ========================
+//
+// We need to be able to deal with many types of input plug, which
+// we do by calling generic lambdas after downcasting to the specific
+// type we are dealing with.
+
+// If `plug` is a supported type, downcasts it to its true type and
+// calls `functor( plug, args )`. Otherwise, does nothing.
+template<typename F>
+void dispatchPlugFunction( const ValuePlug *plug, F &&functor )
```

Comment:
do you think this might be useful elsewhere like IECore::dispatch? Perhaps this could be in PlugAlgo or similar

Reply:
Yeah, I think we might end up with something like that at some point - I did something similar in RandomChoice as well. A generic PlugAlgo function would be slightly different in that it would need to support all plug types though, and then Collect and RandomChoice would need to be adapted to filter out the ones they don't support. I expect we'll do that at some point, but I don't see any harm in waiting for another use case or two to see if it changes the requirements.